soc/nordic/common/dmm.c

```diff
@@ -0,0 +1,272 @@
+DT_MEMORY_REGION_FOREACH_STATUS_OKAY_NODE(_DECLARE_LINKER_VARS);
+
+struct dmm_region {
+	uintptr_t dt_addr;
+	size_t dt_size;
+	uint32_t dt_attr;
+	void *dt_allc;
+};
+
+struct dmm_heap {
+	struct sys_heap heap;
+	const struct dmm_region *region;
+};
+
+static const struct dmm_region dmm_regions[] = {
+	DT_MEMORY_REGION_FOREACH_STATUS_OKAY_NODE(_BUILD_MEM_REGION)};
+
+struct {
+	struct dmm_heap dmm_heaps[ARRAY_SIZE(dmm_regions)];
+} dmm_heaps_data;
+
+static struct dmm_heap *dmm_heap_find(uintptr_t region_addr)
+{
+	struct dmm_heap *dh;
+	for (size_t idx = 0; idx < ARRAY_SIZE(dmm_heaps_data.dmm_heaps); idx++) {
+		dh = &dmm_heaps_data.dmm_heaps[idx];
+		if (dh->region->dt_addr == region_addr) {
+			return dh;
+		}
+	}
+
+	return NULL;
+}
+
+static bool dmm_is_cacheable(const struct dmm_region *region)
+{
+	return (IS_ENABLED(CONFIG_DCACHE) && !!(region->dt_attr & DT_MEM_CACHEABLE));
+}
+
+static bool dmm_is_contained_in_region(size_t start, size_t size, size_t reg_start, size_t reg_size)
+{
+	return ((start >= reg_start) && ((start + size) <= (reg_start + reg_size)));
+}
+
+static size_t dmm_heap_start_get(struct dmm_heap *dh)
+{
+	return WB_UP(dh->region->dt_allc);
+}
+
+static size_t dmm_heap_size_get(struct dmm_heap *dh)
+{
+	return (dh->region->dt_size - (dmm_heap_start_get(dh) - dh->region->dt_addr));
+}
+
+static bool dmm_user_buffer_alloc_check(void *user_buffer, size_t user_length,
+					const struct dmm_region *region)
+{
+	uintptr_t addr = (uintptr_t)user_buffer;
+	size_t buffer_end = addr + user_length;
+
+	/* Check if the buffer is entirely contained in the region */
+	if (dmm_is_contained_in_region(addr, user_length, region->dt_addr, region->dt_size)) {
+		/* Check if the buffer is in cacheable region */
+		if (dmm_is_cacheable(region)) {
+			/* Check if the buffer is aligned and padded to the cache line */
+			if (IS_ALIGNED(addr, DMM_DCACHE_LINE_SIZE) &&
+			    IS_ALIGNED(buffer_end, DMM_DCACHE_LINE_SIZE)) {
+				return true;
+			}
+			return false;
+		}
+		return true;
+	}
+	return false;
+}
+
+static void *dmm_buffer_alloc(struct dmm_heap *dh, size_t length)
+{
+	length = ROUND_UP(length, DMM_DCACHE_LINE_SIZE);
+	return sys_heap_aligned_alloc(&dh->heap, DMM_DCACHE_LINE_SIZE, length);
+}
+
+static void dmm_buffer_free(struct dmm_heap *dh, void *buffer)
+{
+	sys_heap_free(&dh->heap, buffer);
+}
+
+int dmm_buffer_out_prepare(void *region, void *user_buffer, size_t user_length, void **buffer_out)
+{
+	struct dmm_heap *dh;
+
+	// Get memory region that specified device can perform DMA transfers from
+	dh = dmm_heap_find((uintptr_t)region);
+	if (!dh) {
+		return -EINVAL;
+	}
+
+	// Check if:
+	// - provided user buffer is already in correct memory region,
+	// - provided user buffer is aligned and padded to cache line, if it is located in cacheable
+	// region.
+	if (dmm_user_buffer_alloc_check(user_buffer, user_length, dh->region)) {
+		// If yes, assign buffer_out to user_buffer
+		*buffer_out = user_buffer;
```

Comment:
In PSA many of the user buffers are const and read only so this code creates a problem.